### PR TITLE
Upgrade Rust to 1.81

### PR DESCRIPTION
**Pull request by Schaeff**
This is a prerequisite for #1845 

TODO:
- [x] narrow down or remove the dead_code exception

### CODE REVIEW

File: backend/src/field_filter.rs [hunk 1]
```diff
@@ -34,10 +34,10 @@ macro_rules! generalize_factory {
                             unsafe {
                                 let result = <$restricted_factory as BackendFactory<$supported_type>>::create(
                                     &$restricted_factory,
-                                    std::mem::transmute(pil), std::mem::transmute(fixed), output_dir, setup,
+                                    std::mem::transmute::<std::sync::Arc<powdr_ast::analyzed::Analyzed<F>>, std::sync::Arc<powdr_ast::analyzed::Analyzed<$supported_type>>>(pil), std::mem::transmute::<std::sync::Arc<std::vec::Vec<(std::string::String, powdr_executor::constant_evaluator::VariablySizedColumn<F>)>>, std::sync::Arc<std::vec::Vec<(std::string::String, powdr_executor::constant_evaluator::VariablySizedColumn<$supported_type>)>>>(fixed), output_dir, setup,
```
Comment (leonardoalt) [inline backend/src/field_filter.rs:37]: 1) what

Comment (chriseth) [reply backend/src/field_filter.rs:37]: Wait this evil macro was merged after all?

File: riscv/src/asm/mod.rs [hunk 2]
```diff
@@ -1,3 +1,8 @@
+// this is required to please clippy which finds dome dead code in the
```
Comment (chriseth) [inline riscv/src/asm/mod.rs:1]: But the generated parser already has `allow(all)` (in parser.rs) - where exactly is the dead code?

Comment (Schaeff) [reply riscv/src/asm/mod.rs]: The parser has `allow(clippy::all)` but `dead_code` is a rustc lint not a clippy lint. I moved it there!

File: backend/src/field_filter.rs (hunk 1 above)
Comment (lvella) [reply backend/src/field_filter.rs]: > Wait this evil macro was merged after all?

The first `unsafe` solution got nuked by rust. The evil macro was the only PR in town.
https://github.com/rust-lang/rust/pull/120248#issuecomment-2340261183

**Review by chriseth — APPROVED** [review]

**Pull request merged**